File: python/0104二叉树的最大深度.py
from binarytree import TreeNode,pre_create


# 递归
def maxDepth(root:TreeNode) -> int:
    if(not root):
        return 0
    
    return 1+max(maxDepth(root.left),maxDepth(root.right))
def maxDepth(root:TreeNode) -> int:
    
    depth=0
    
    # 前序遍历
    def _get_depth(root:TreeNode,level:int):
        # 闭包中访问外部变量关键字nonelocal
        nonlocal depth
        if(root):
            depth=level if level>depth else depth
            _get_depth(root.left,level+1)
            _get_depth(root.right,level+1)
    
    _get_depth(root,1)
    return depth

print(maxDepth(pre_create([1,2,3])))

File: cpp/0104二叉树的最大深度.cpp
#include<deque>
#include"mybitree.h"

// 迭代
int maxDepth(TreeNode *root)
{
    int res = 0;
    deque<TreeNode *> nodes;
    nodes.push_back(root);
    while(!nodes.empty())
    {
        bool exist_node = false;
        int size = nodes.size();
        for (int i = 0;i<size;++i)
        {
            TreeNode *crt = nodes.front();
            if (crt)
            {
                exist_node = true;
                nodes.push_back(crt->left);
                nodes.push_back(crt->right);
            }
            nodes.pop_front();
        }
        res += exist_node;
    }

    return res;
}